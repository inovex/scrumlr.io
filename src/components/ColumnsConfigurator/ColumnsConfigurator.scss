@import "constants/style";

.columns-configurator {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;

  // allow horizontal scroll
  overflow: auto hidden;
  width: 100%;
  max-width: 100%;
  height: 100%;

  @include scrollbar();
}

.columns-configurator__columns-wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;
}
