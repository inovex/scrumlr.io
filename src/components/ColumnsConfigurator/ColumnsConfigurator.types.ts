export type ColumnPlacement = "first" | "center" | "last" | "all";

export type AddTemplateColumnAlignment = "left" | "right";
