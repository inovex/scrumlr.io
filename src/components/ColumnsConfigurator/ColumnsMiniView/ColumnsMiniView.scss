@import "constants/style";

// magic w/h
$mini-column-width: 46px;
$mini-column-height: 91px;

.columns-mini-view {
  display: flex;
  flex-direction: row;
  align-items: center;

  // allow horizontal scroll
  overflow: auto hidden;
  width: 100%;
  max-width: 100%;
  height: 100%;

  scrollbar-width: none; // TODO not sure whether to hide, will decide later
}

.columns-mini-view__column {
  display: flex;
  justify-content: center;
  align-items: center;

  width: $mini-column-width;
  min-width: $mini-column-width;
  height: $mini-column-height;

  border: 2px solid var(--accent-color--light);

  // deactivate borders depending on position
  // TODO maybe different borders
  &--border {
    &-first {
      border-right: none;
    }
    &-center {
      border-left: none;
      border-right: none;
    }
    &-last {
      border-left: none;
    }
  }

  &--hidden {
    // TODO thinner stripes
    background-image: $column-stripes--light;
  }
}

.columns-mini-view__column-index {
  color: $navy--400;
  font-size: $text--base;
  font-weight: 600;
}

[theme="dark"] {
  .columns-mini-view__column {
    &--hidden {
      background-image: $column-stripes--dark;
    }
  }

  .columns-mini-view__column-index {
    color: $gray--000;
  }
}
