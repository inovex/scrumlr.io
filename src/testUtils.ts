export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;