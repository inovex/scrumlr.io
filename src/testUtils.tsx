import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
