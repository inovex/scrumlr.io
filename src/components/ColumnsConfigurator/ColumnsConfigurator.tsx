import {uniqueId} from "underscore";
import classNames from "classnames";
import {useState} from "react";
import {EditableTemplateColumn, TemplateColumn} from "store/features";
import {Color, getColorClassName, getNextColor, getPreviousColor} from "constants/colors";
import {closestCenter, DndContext, DragEndEvent, DragOverEvent, DragOverlay, DragStartEvent, PointerSensor, useSensor, useSensors} from "@dnd-kit/core";
import {horizontalListSortingStrategy, SortableContext} from "@dnd-kit/sortable";
import {ColumnsConfiguratorColumn} from "./ColumnsConfiguratorColumn/ColumnsConfiguratorColumn";
import {AddTemplateColumn} from "./AddTemplateColumn/AddTemplateColumn";
import {calcPlacement} from "./ColumnsConfigurator.utils";
import {AddTemplateColumnAlignment} from "./ColumnsConfigurator.types";
import "./ColumnsConfigurator.scss";

export type ColumnsConfiguratorProps = {
  className: string;
  templateId: string;
  columns: EditableTemplateColumn[]; // without deleted cols
  addColumn: (templateColumn: TemplateColumn, index: number) => void;
  moveColumn: (fromIndex: number, toIndex: number) => void;
  editColumn: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void;
  deleteColumn: (templateColumn: EditableTemplateColumn) => void;
};

export const ColumnsConfigurator = (props: ColumnsConfiguratorProps) => {
  // id of column which is actively being dragged
  const [activeElementId, setActiveElementId] = useState<string | null>(null);
  const [activeColumn, setActiveColumn] = useState<EditableTemplateColumn | null>(null);
  const [dropElementId, setDropElementId] = useState<string | null>(null);

  const sensors = useSensors(useSensor(PointerSensor));

  const handleDragStart = (event: DragStartEvent) => {
    const activeId = event.active.id as string;
    setActiveElementId(activeId);
    setActiveColumn(props.columns.find((c) => c.id === activeId)!);
  };

  const handleDragOver = (event: DragOverEvent) => {
    setDropElementId((event.over?.id as string) ?? null);
  };

  const handleDragEnd = (event: DragEndEvent) => {
    const {active, over} = event;

    if (active.id !== over?.id) {
      const oldIndex = props.columns.findIndex((c) => c.id === active.id);
      const newIndex = props.columns.findIndex((c) => c.id === over?.id);

      props.moveColumn(oldIndex, newIndex);
    }

    // don't forget to clear states
    setActiveElementId(null);
    setActiveColumn(null);
    setDropElementId(null);
  };

  // function to calculate where each column would be positioned if the drag-and-drop operation were completed at its current state
  const getPotentialIndex = (index: number) => {
    // no valid drag operation occurring
    if (!activeElementId || !dropElementId) return index;

    const dragElementIndex = props.columns.findIndex((c) => c.id === activeElementId);
    const dropElementIndex = props.columns.findIndex((c) => c.id === dropElementId);

    // if current index matches active index of drag element, the column is being dragged there,
    // so the potential index is that of the drop element
    if (index === dragElementIndex) {
      return dropElementIndex;
    }

    // dragging right: active index less the drop zone, and element is between them, the element shifts to the left
    if (dragElementIndex < dropElementIndex && index > dragElementIndex && index <= dropElementIndex) {
      return index - 1;
    }

    // dragging left: analogue to above, element shifts to the right
    if (dragElementIndex > dropElementIndex && index < dragElementIndex && index >= dropElementIndex) {
      return index + 1;
    }

    // otherwise the index remains unchanged
    return index;
  };

  // renders the drag overlay, which is the element which is dragged
  const renderDragOverlay = () => {
    if (!activeElementId || !activeColumn) return null;
    const activeColumnIndex = getPotentialIndex(props.columns.findIndex((c) => c.id === activeColumn.id));

    return (
      <ColumnsConfiguratorColumn
        className={classNames("columns-configurator__column", "columns-configurator__column--ghost")}
        column={activeColumn}
        // here, use actual index unlike potential index in the main part
        index={activeColumnIndex}
        placement={calcPlacement(activeColumnIndex, props.columns)}
        allColumns={props.columns}
      />
    );
  };

  const addTemplateColumn = (alignment: AddTemplateColumnAlignment, color: Color) => {
    const newColumn: TemplateColumn = {
      id: uniqueId("col"),
      template: props.templateId,
      name: `Column ${props.columns.length + 1}`,
      description: "",
      color,
      visible: true,
      index: -1, // will be overwritten anyway
    };

    // target index for the right has to take deleted cols which are still part of the array into account,
    // which is why we use the total col amount as new index
    const index = alignment === "left" ? 0 : props.columns.length;

    props.addColumn(newColumn, index);
  };

  if (props.columns.length === 0) {
    return null;
  }

  return (
    <DndContext sensors={sensors} collisionDetection={closestCenter} onDragStart={handleDragStart} onDragOver={handleDragOver} onDragEnd={handleDragEnd}>
      <SortableContext items={props.columns} strategy={horizontalListSortingStrategy}>
        <div className={classNames(props.className, "columns-configurator", getColorClassName("backlog-blue"))}>
          <AddTemplateColumn alignment="left" color={getPreviousColor(props.columns[0].color)} onClick={addTemplateColumn} />
          <div className="columns-configurator__columns-wrapper">
            {props.columns.map((column, index) => {
              const potentialIndex = getPotentialIndex(index);
              return (
                <ColumnsConfiguratorColumn
                  className="columns-configurator__column"
                  key={column.id}
                  column={column}
                  index={potentialIndex}
                  activeDrag={column.id === activeElementId}
                  activeDrop={column.id === dropElementId}
                  placement={calcPlacement(potentialIndex, props.columns)}
                  allColumns={props.columns}
                  editColumn={props.editColumn}
                  deleteColumn={props.deleteColumn}
                />
              );
            })}
          </div>
          <AddTemplateColumn alignment="right" color={getNextColor(props.columns[props.columns.length - 1].color)} onClick={addTemplateColumn} />
        </div>
      </SortableContext>

      <DragOverlay>{renderDragOverlay()}</DragOverlay>
    </DndContext>
  );
};
