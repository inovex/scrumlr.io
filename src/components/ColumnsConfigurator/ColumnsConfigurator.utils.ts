import {EditableTemplateColumn, ExistenceStatus, Template, TemplateColumn, TemplateWithColumns} from "store/features";
import {ColumnPlacement} from "./ColumnsConfigurator.types";

export const calcPlacement = (index: number, columns: TemplateColumn[]): ColumnPlacement => {
  const {length} = columns;
  if (length === 1) {
    return "all";
  }
  if (index === 0) {
    return "first";
  }
  if (index === length - 1) {
    return "last";
  }
  return "center";
};

// convert TemplateColumn to EditableColumn, used for test purposes
export const convertToEditableColumn = (c: TemplateColumn, overrideExistenceStatus?: Partial<ExistenceStatus>): EditableTemplateColumn =>
  ({
    ...c,
    persisted: false,
    mode: undefined,
    ...overrideExistenceStatus,
  }) as EditableTemplateColumn;

// from an array of all templates and column, create a TemplateAndColumns array with all corresponding pairs
export const mergeTemplateAndColumns = (templates: Template[], templateColumns: TemplateColumn[]) =>
  templates.map((template) => ({
    template,
    columns: templateColumns.filter((column) => column.template === template.id),
  })) as TemplateWithColumns[];

export const getTemplateAndColumnsByTemplateId = (templatesAndColumns: {templates: Template[]; templateColumns: TemplateColumn[]}, templateId: string) =>
  mergeTemplateAndColumns(templatesAndColumns.templates, templatesAndColumns.templateColumns).find((tac) => tac.template.id === templateId);
