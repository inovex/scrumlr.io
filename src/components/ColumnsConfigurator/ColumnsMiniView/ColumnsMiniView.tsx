import {EditableTemplateColumn} from "store/features";
import classNames from "classnames";
import {getColorClassName} from "constants/colors";
import {calcPlacement} from "../ColumnsConfigurator.utils";
import "./ColumnsMiniView.scss";

type ColumnsMiniViewProps = {
  className: string;
  columns: EditableTemplateColumn[];
};

export const ColumnsMiniView = (props: ColumnsMiniViewProps) => (
    <div className={classNames(props.className, "columns-mini-view")}>
      {props.columns.map((column) => (
        <div
          key={column.id}
          className={classNames(
            "columns-mini-view__column",
            {"columns-mini-view__column--hidden": !column.visible},
            `columns-mini-view__column--border-${calcPlacement(column.index, props.columns)}`,
            getColorClassName(column.color)
          )}
        >
          <span className="columns-mini-view__column-index">{column.index + 1}</span>
        </div>
      ))}
    </div>
  );
