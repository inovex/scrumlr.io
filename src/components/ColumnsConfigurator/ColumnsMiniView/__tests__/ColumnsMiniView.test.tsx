import {ColumnsMiniView, ColumnsMiniViewProps} from "../ColumnsMiniView";
import getTestApplicationState from "utils/test/getTestApplicationState";
import {convertToEditableColumn} from "../../ColumnsConfigurator.utils";
import {render} from "testUtils";

const renderColumnsMiniView = (templateId: string) => {
  const columns = getTestApplicationState()
    .templatesColumns.filter((tc) => tc.template === templateId)
    .map((tc) => convertToEditableColumn(tc));

  return render(<ColumnsMiniView className={"test-columns-mini-view"} columns={columns} />);
};

describe("ColumnsMiniView", () => {
  it("should match render of template 1", () => {
    const {container} = renderColumnsMiniView("test-templates-id-1");
    expect(container).toMatchSnapshot();
    expect(container.firstChild?.childNodes).toHaveLength(2);
  });

  it("should match render of template 2", () => {
    const {container} = renderColumnsMiniView("test-templates-id-2");
    expect(container).toMatchSnapshot();
    expect(container.firstChild?.childNodes).toHaveLength(1);
  });
});
