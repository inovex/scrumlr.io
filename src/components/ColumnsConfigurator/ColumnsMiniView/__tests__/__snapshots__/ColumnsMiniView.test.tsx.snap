// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnsMiniView should match render of template 1 1`] = `
<div>
  <div
    class="test-columns-mini-view columns-mini-view"
  >
    <div
      class="columns-mini-view__column columns-mini-view__column--border-first accent-color__backlog-blue"
    >
      <span
        class="columns-mini-view__column-index"
      >
        1
      </span>
    </div>
    <div
      class="columns-mini-view__column columns-mini-view__column--hidden columns-mini-view__column--border-last accent-color__planning-pink"
    >
      <span
        class="columns-mini-view__column-index"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`ColumnsMiniView should match render of template 2 1`] = `
<div>
  <div
    class="test-columns-mini-view columns-mini-view"
  >
    <div
      class="columns-mini-view__column columns-mini-view__column--border-all accent-color__goal-green"
    >
      <span
        class="columns-mini-view__column-index"
      >
        1
      </span>
    </div>
  </div>
</div>
`;
